webgoat-lessons/xxe/src/main/resources/lessonPlans/en/XXE_changing_content_type_solution.adoc

```diff
@@ -0,0 +1,71 @@
+=== Assignment solution
+
+The idea behind this assignment is that while it may look that the application is only accepting JSON but if we change the body of the message to XML the framework might process it. When you try entering a comment the body of request will be:
+
+[source, json]
+----
+{"text":"My first comment"}
+----
+
+This is a normal json message, let's try to change the content-type of the request
+
+[source]
+----
+POST http://localhost:8080/WebGoat/xxe/content-type HTTP/1.1
+Content-Type: application/xml
+
+{"text":"My first comment"}
+----
+
+this results in the following exception:
+
+[source]
+----
+javax.xml.bind.UnmarshalException\n - with linked exception:\n[javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]\nMessage: Content is not allowed in prolog.
+----
+
+Depending on the XML parser you might get a better error message, in this case the message is a bit cryptic, it means that we are not sending valid xml. For example the Jackson library gives the following message:
+
+[source]
+----
+JSON parse error: Unexpected character '{' (code 123) in prolog; expected
+      '<'\n at [row,col {unknown-source}]: [1,1]; nested exception is com.fasterxml.jackson.core.JsonParseException:
+      Unexpected character '{' (code 123) in prolog; expected '<'\n at [row,col {unknown-source}]: [1,1]“
+----
+
+This error message appears because we are still sending a json message towards the endpoint, so if we intercept and change change the json message to a xml message:
+
+[souce]
+----
+POST http://localhost:8080/WebGoat/xxe/content-type HTTP/1.1
+Content-Type: application/xml
+
+<text>This is my first message</text>
+----
+
+Again an error message from the endpoint:
+
+[source]
+----
+"javax.xml.bind.UnmarshalException\\n - with linked exception:\\n[com.sun.istack.SAXParseException2; lineNumber: 1; columnNumber: 7; unexpected element (uri:\\\"\\\", local:\\\"text\\\"). Expected elements are <{}comment>]
+----
+
+The parser complains that the message is not a valid xml message and needs to be embedded in a `comment` tag:
+
+[source, xml]
+----
+POST http://localhost:8080/WebGoat/xxe/content-type HTTP/1.1
+Content-Type: application/xml
+
+<comment><text>This is my first message</text></comment>
+----
+
+The endpoint no longer complains and if you refresh the page in WebGoat the posted comments appear. For the attack to work we need to post:
```

Comment:
Perhaps here is a good place to mention that on some company networks some network devices might drop this payload completely if it was sent over HTTP. In that case no response is returned on the POST and the WebGoat never receives the request. However these kind of protections are of limited use, as the same request would pass successfully in an HTTPS setup, where that payload would be encrypted.

Reply:
Good point, added